File: tests/python/unit/test_unit_complex_mixer.py
import numpy as np
import pytest
from complex_mixer import ComplexMixer

# Define the test object.
@pytest.fixture
def complex_mixer():
    return ComplexMixer()

def test_complex_mixer_valid_output(complex_mixer):
    # Define the input.
    f_shift = 150.00
    fs = 1000.00
    f = 300.00
    t = np.linspace(0, 1, int(fs), endpoint=False)
    I = np.cos(2 * np.pi * f * t)
    Q = np.zeros_like(t)

    # Compute the result.
    I_out = complex_mixer(f_shift, fs, I, Q)
    I_out = np.abs(np.fft.fft(I_out))

    # Test the result.
    spec = [int(f - f_shift), int(f + f_shift)]

    assert(I_out.size == I.size), "Invalid size of the shifted signal."
    assert(np.allclose(I_out[spec], fs / 4, atol=5e-2)), "Spectral components must be at f - f_shift and f + f_shift."
    assert(np.allclose(np.delete(I_out[:I_out.size//2], spec), 0, atol=5e-2)), "Noise must be close to 0."

def test_complex_mixer_invalid_input(complex_mixer):
    with pytest.raises(ValueError, match="ComplexMixer.__call__: Sampling frequency must be positive."):
        complex_mixer(200.0, -1000, np.array([1]), np.array([1]))

File: tests/cpp/unit/testUnitComplexMixer.cpp
#include <algorithm>
#include <cmath>
#include <gtest/gtest.h>
#include "utils.h"
#include "dsp.h"
#include "complexMixer.h"

using namespace std;

// Define the test object.
class TestComplexMixer : public ::testing::Test {
    protected:
        dsp::ComplexFFT fft;
        ComplexMixer complexMixer;
};

TEST_F(TestComplexMixer, validOutput) {
    // Define the input.
    double fshift = 150.0;
    double fs = 1000.0;
    double f = 300.0;
    vector<double> I = utils::linspace(0.0, 1.0, static_cast<uint>(fs), false);
    for (uint i = 0; i < I.size(); i++) { I[i] = cos(2 * M_PI * f * I[i]); }
    vector<double> Q(static_cast<uint>(fs), 0);

    // Compute the result.
    vector<double> Iout = complexMixer(fshift, fs, I, Q);
    vector<complex<double>> fftIout = fft.fft(Iout);

    // Test the result.
    vector<uint> spec = {
        static_cast<uint>(f - fshift),
        static_cast<uint>(f + fshift)
    };

    ASSERT_EQ(fftIout.size(), I.size()) << "Invalid size of the shifted signal.";
    for (uint i = 0; i < fftIout.size() / 2; i++) {
        if (find(spec.begin(), spec.end(), i) != spec.end()) {
            ASSERT_NEAR(abs(fftIout[i]), fs / 4, 5e-2) << "Spectral components must be at f - fshift and f + fshift.";
        }
        else { ASSERT_NEAR(abs(fftIout[i]), 0, 5e-2) << "Noise must be close to 0."; }
    }
}

TEST_F(TestComplexMixer, invalidInput) {
    double fs = -1000.0; // Sampling frequency must be positive.
    EXPECT_THROW(complexMixer(200.0, fs, {1}, {1}), invalid_argument);
}
